rfd/0197-better-async-state-management-with-tanstack-query.md

```diff
@@ -0,0 +1,266 @@
+
+### useAttempt
+
+This is a custom hook that is used to manage any state in Teleport. The attempt can have different statuses (none,
+loading, error, success).
+
+This results in a lot of boilerplate code in managing the data fetching & state storage, and error handling.
+
+#### Example query
+
+```tsx
+function fetchAccessList() {
+  scopedAttempt.setAttempt({ status: 'processing' });
+
+  accessManagementService
+    .fetchAccessList(accessListId)
+    .then(fetchedAccessList => {
+      const [modifiedAccessList, newPerms] = modifyAccessList(
+        fetchedAccessList,
+        accessLists,
+        ctx
+      );
+      setAccessList(modifiedAccessList);
+      setPerms(newPerms);
+      scopedAttempt.setAttempt({ status: 'success' });
+    })
+    .catch((e: Error) =>
+      scopedAttempt.setAttempt({ status: 'failed', statusText: e.message })
+    );
+}
+```
+
+This then has to be called with an effect that watches the access list ID:
+
+```tsx
+// The accessListId can change if a user clicks on a different
+// access list in the notification dropdown.
+useEffect(() => {
+  if (!accessList || accessList.id !== accessListId) {
+    fetchAccessList();
+  }
+}, [accessListId]);
+```
+
+#### Example mutation
+
+```tsx
+const { attempt, setAttempt } = useAttempt();
+const isDisabled = attempt.status === 'processing';
+
+function onOk() {
+  setAttempt({ status: 'processing' });
+  accessManagementService
+    .deleteAccessList(accessListId)
+    .then(() =>
+      history.push(cfg.getAccessListManagementRoute(), {
+        deletedAccessListId: accessListId,
+      })
+    )
+    .catch((e: Error) =>
+      setAttempt({ status: 'failed', statusText: e.message })
+    );
+}
+```
+
+### useAsync
+
+This is a newer custom hook that is used to manage async state in Teleport. It requires less boilerplate than
+`useAttempt`, as it handles running async operations and handling errors, but still requires a manual trigger in an
+effect,
+which causes data to not fetch until at least the component has rendered once.
+
+Whilst this works fine, it lacks an extended API that TanStack Query can provide.
+
+#### Example query
+
+```tsx
+const [fetchQrCodeAttempt, fetchQrCode] = useAsync((privilegeToken: string) =>
+  auth.createMfaRegistrationChallenge(privilegeToken, 'totp')
+);
+
+useEffect(() => {
+  fetchQrCode(privilegeToken);
+}, []);
+```
+
+#### Example mutation
+
+```tsx
+const [deleteRequestAttempt, runDeleteRequest] = useAsync(async () => {
+  await ctx.workflowService.deleteAccessRequest(props.requestId);
+  historyService.replace(cfg.getAccessRequestRoute());
+});
+```
+
+## TanStack Query
+
+TanStack Query offers many powerful abilities, including:
+
+- Automatic retries and manual re-fetching methods
+  - Automatic retries can be configured to retry failed requests a certain number of times, with exponential backoff, something not easily achievable
+    with `useAttempt` or `useAsync`
```

Comment:
i can't remember on top of my head what it exactly it was, but there was a case where @flyinghermit ran into backend caching, where the only solution was to retry in a for loop after some timeout for maximum of some amount. i think we were looping until we got a specific json field back in the response?, perhaps tanstack can handle that for us too?

Reply:
I think if the `queryFn` was to throw an error if that field wasn't there, TanStack Query would see it as an error and retry (but also set the loading state to an error)

```ts
const query = useQuery({
  queryKey: ['some', 'query', 'key'],
  queryFn: ({ signal }) => {
    const res = await service.fetchSomething(signal);

    if (!res.whatever) {
      throw new Error('does not exist');
    }

    return res;
  },
});
````

Or maybe something in an effect

```ts
const query = useQuery({
  queryKey: ['some', 'query', 'key'],
  queryFn: ({ signal }) => service.fetchSomething(signal),
});

useEffect(() => {
  if (!query.isSuccess || !query.data.whatever) {
    return;
  }

  void query.refetch(); // maybe a timeout here
}, [query.isSuccess, query.refetch]);
```